// @huggingface/blake3-jit/dist/commonjs/hasher.js
"use strict";
/**
 * BLAKE3 Hasher - Incremental hashing with support for all modes
 *
 * Supports:
 * - Regular hashing
 * - Keyed hashing (MAC)
 * - Key derivation (derive_key)
 * - XOF (eXtendable Output Function) mode
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Hasher = exports.XofReader = void 0;
const compress_js_1 = require("./compress.js");
const constants_js_1 = require("./constants.js");
const utils_js_1 = require("./utils.js");
/**
 * Output state for XOF (eXtendable Output Function) mode.
 * Allows reading arbitrary amounts of output.
 */
class XofReader {
    inputCv;
    blockWords;
    counter;
    blockLen;
    flags;
    outputBlock;
    outputBlockOffset;
    constructor(inputCv, blockWords, counter, blockLen, flags) {
        this.inputCv = inputCv;
        this.blockWords = blockWords;
        this.counter = counter;
        this.blockLen = blockLen;
        this.flags = flags | constants_js_1.ROOT;
        this.outputBlock = new Uint32Array(16);
        this.outputBlockOffset = 64; // Forces generation on first read
    }
    /**
     * Read the next `length` bytes of output.
     */
    read(length) {
        const output = new Uint8Array(length);
        let outputOffset = 0;
        while (outputOffset < length) {
            // Generate new output block if needed
            if (this.outputBlockOffset >= 64) {
                (0, compress_js_1.compress)(this.inputCv, 0, this.blockWords, 0, this.outputBlock, 0, true, // full 64-byte output
                this.counter++, this.blockLen, this.flags);
                this.outputBlockOffset = 0;
            }
            // Copy bytes from output block
            const available = 64 - this.outputBlockOffset;
            const toCopy = Math.min(available, length - outputOffset);
            // Optimized copy using writeLittleEndianBytesPartial
            const wordOffset = this.outputBlockOffset >>> 2;
            const byteWithinWord = this.outputBlockOffset & 3;
            if (byteWithinWord === 0 && toCopy >= 4) {
                // Aligned copy - can use word-at-a-time
                const fullWords = toCopy >>> 2;
                (0, utils_js_1.writeLittleEndianBytesPartial)(this.outputBlock, wordOffset, output, outputOffset, fullWords << 2);
                const bytesCopied = fullWords << 2;
                outputOffset += bytesCopied;
                this.outputBlockOffset += bytesCopied;
            }
            else {
                // Byte-by-byte for unaligned access
                for (let i = 0; i < toCopy; i++) {
                    const wordIdx = (this.outputBlockOffset + i) >>> 2;
                    const byteIdx = (this.outputBlockOffset + i) & 3;
                    output[outputOffset + i] = (this.outputBlock[wordIdx] >>> (byteIdx << 3)) & 0xff;
                }
                outputOffset += toCopy;
                this.outputBlockOffset += toCopy;
            }
        }
        return output;
    }
}
exports.XofReader = XofReader;
/**
 * Chunk state for processing input data.
 * Each chunk is 1024 bytes and produces an 8-word chaining value.
 */
class ChunkState {
    chainingValue;
    chunkCounter;
    blockWords;
    blockLen;
    blocksCompressed;
    flags;
    constructor(keyWords, chunkCounter, flags) {
        this.chainingValue = new Uint32Array(keyWords);
        this.chunkCounter = chunkCounter;
        this.blockWords = new Uint32Array(16);
        this.blockLen = 0;
        this.blocksCompressed = 0;
        this.flags = flags;
    }
    resetTo(keyWords, chunkCounter, flags) {
        this.chainingValue.set(keyWords);
        this.chunkCounter = chunkCounter;
        this.blockLen = 0;
        this.blocksCompressed = 0;
        this.flags = flags;
    }
    /**
     * Get the flags for the current block.
     */
    startFlag() {
        return this.blocksCompressed === 0 ? constants_js_1.CHUNK_START : 0;
    }
    /**
     * Update the chunk state with input data.
     * Returns the number of bytes consumed.
     */
    update(input, inputOffset, inputLen) {
        let consumed = 0;
        while (inputLen > 0) {
            // If we have a full block, compress it
            if (this.blockLen === constants_js_1.BLOCK_LEN) {
                (0, compress_js_1.compress)(this.chainingValue, 0, this.blockWords, 0, this.chainingValue, 0, false, this.chunkCounter, constants_js_1.BLOCK_LEN, this.flags | this.startFlag());
                this.blocksCompressed++;
                this.blockLen = 0;
            }
            // Fill the block buffer
            const want = constants_js_1.BLOCK_LEN - this.blockLen;
            const take = Math.min(want, inputLen);
            if (this.blockLen === 0 && take === constants_js_1.BLOCK_LEN) {
                (0, utils_js_1.readLittleEndianWordsFull)(input, inputOffset, this.blockWords);
            }
            else {
                // Partial block - byte-by-byte into correct position
                for (let i = 0; i < take; i++) {
                    const pos = this.blockLen + i;
                    const wordIdx = pos >>> 2;
                    const byteIdx = pos & 3;
                    if (byteIdx === 0) {
                        this.blockWords[wordIdx] = input[inputOffset + i];
                    }
                    else {
                        this.blockWords[wordIdx] |= input[inputOffset + i] << (byteIdx << 3);
                    }
                }
            }
            this.blockLen += take;
            inputOffset += take;
            inputLen -= take;
            consumed += take;
        }
        return consumed;
    }
    /**
     * Finalize this chunk and return its output.
     * Returns 8 words (chaining value) or 16 words (if root).
     */
    output() {
        // Zero-pad unused words in blockWords to avoid stale data from previous blocks
        // This is necessary when a partial block follows a full block within the same chunk
        const usedWords = (this.blockLen + 3) >>> 2; // ceil(blockLen / 4)
        for (let i = usedWords; i < 16; i++) {
            this.blockWords[i] = 0;
        }
        return {
            inputCv: this.chainingValue,
            blockWords: this.blockWords,
            blockLen: this.blockLen,
            counter: this.chunkCounter,
            flags: this.flags | this.startFlag() | constants_js_1.CHUNK_END,
        };
    }
    /**
     * Get the number of bytes in this chunk.
     */
    len() {
        return this.blocksCompressed * constants_js_1.BLOCK_LEN + this.blockLen;
    }
}
/**
 * Main BLAKE3 Hasher class.
 *
 * Usage:
 *   const hasher = new Hasher();
 *   hasher.update(data);
 *   const hash = hasher.finalize();
 *
 * Or with chaining:
 *   const hash = new Hasher().update(data).finalize();
 */
class Hasher {
    chunkState;
    keyWords;
    cvStack;
    cvStackLen;
    flags;
    parentBlock;
    parentCv;
    chunkCv;
    outWords;
    finalizeCv;
    /**
     * Create a new Hasher.
     *
     * @param keyWords - Initial key words (IV for regular hashing)
     * @param flags - Domain separation flags
     */
    constructor(keyWords, flags) {
        this.keyWords = keyWords ? new Uint32Array(keyWords) : new Uint32Array(constants_js_1.IV);
        this.flags = flags ?? 0;
        this.chunkState = new ChunkState(this.keyWords, 0, this.flags);
        this.cvStack = new Uint32Array(constants_js_1.MAX_DEPTH * 8);
        this.cvStackLen = 0;
        this.parentBlock = new Uint32Array(16);
        this.parentCv = new Uint32Array(8);
        this.chunkCv = new Uint32Array(8);
        this.outWords = new Uint32Array(16);
        this.finalizeCv = new Uint32Array(8);
    }
    /**
     * Reset the hasher to process a new message with the same key/flags.
     * Reuses all internal buffers — zero allocations.
     */
    reset() {
        this.chunkState.resetTo(this.keyWords, 0, this.flags);
        this.cvStackLen = 0;
        return this;
    }
    /**
     * Create a new keyed hasher (MAC).
     *
     * @param key - 32-byte key
     */
    static newKeyed(key) {
        if (key.length !== constants_js_1.KEY_LEN) {
            throw new Error(`Key must be ${constants_js_1.KEY_LEN} bytes, got ${key.length}`);
        }
        const keyWords = new Uint32Array(8);
        if (utils_js_1.IS_LITTLE_ENDIAN) {
            const view = new Uint32Array(key.buffer, key.byteOffset, 8);
            keyWords.set(view);
        }
        else {
            for (let i = 0; i < 8; i++) {
                const off = i * 4;
                keyWords[i] = key[off] | (key[off + 1] << 8) | (key[off + 2] << 16) | (key[off + 3] << 24);
            }
        }
        return new Hasher(keyWords, constants_js_1.KEYED_HASH);
    }
    /**
     * Create a new key derivation hasher.
     *
     * @param context - Context string for domain separation
     */
    static newDeriveKey(context) {
        // First, hash the context string with DERIVE_KEY_CONTEXT flag
        const contextBytes = (0, utils_js_1.encodeUTF8)(context);
        const contextHasher = new Hasher(new Uint32Array(constants_js_1.IV), constants_js_1.DERIVE_KEY_CONTEXT);
        contextHasher.update(contextBytes);
        // Get the context key
        const contextKey = new Uint32Array(8);
        const output = contextHasher.finalizeOutput();
        (0, compress_js_1.compress)(output.inputCv, 0, output.blockWords, 0, contextKey, 0, false, output.counter, output.blockLen, output.flags | constants_js_1.ROOT);
        // Return a hasher initialized with the context key
        return new Hasher(contextKey, constants_js_1.DERIVE_KEY_MATERIAL);
    }
    /**
     * Push a chaining value onto the stack.
     */
    pushCv(cv, cvOffset) {
        this.cvStack.set(cv.subarray(cvOffset, cvOffset + 8), this.cvStackLen * 8);
        this.cvStackLen++;
    }
    /**
     * Pop a chaining value from the stack.
     */
    popCv(out, outOffset) {
        this.cvStackLen--;
        out.set(this.cvStack.subarray(this.cvStackLen * 8, (this.cvStackLen + 1) * 8), outOffset);
    }
    /**
     * Add a chunk's chaining value and merge completed subtrees.
     */
    addChunkCv(newCv, newCvOffset, totalChunks) {
        const parentBlock = this.parentBlock;
        const parentCv = this.parentCv;
        while ((totalChunks & 1) === 0) {
            // Pop left child, new CV is right child
            this.popCv(parentBlock, 0);
            parentBlock.set(newCv.subarray(newCvOffset, newCvOffset + 8), 8);
            (0, compress_js_1.compress)(this.keyWords, 0, parentBlock, 0, parentCv, 0, false, 0, constants_js_1.BLOCK_LEN, this.flags | constants_js_1.PARENT);
            newCv = parentCv;
            newCvOffset = 0;
            totalChunks >>>= 1;
        }
        this.pushCv(newCv, newCvOffset);
    }
    /**
     * Update the hasher with input data.
     *
     * @param input - Data to hash
     * @returns this (for chaining)
     */
    update(input) {
        let inputOffset = 0;
        let inputLen = input.length;
        // Fill the current chunk
        while (inputLen > 0) {
            // If current chunk is full, finalize it and start a new one
            if (this.chunkState.len() === constants_js_1.CHUNK_LEN) {
                const output = this.chunkState.output();
                const chunkCv = this.chunkCv;
                (0, compress_js_1.compress)(output.inputCv, 0, output.blockWords, 0, chunkCv, 0, false, output.counter, output.blockLen, output.flags);
                const totalChunks = this.chunkState.chunkCounter + 1;
                this.addChunkCv(chunkCv, 0, totalChunks);
                this.chunkState.resetTo(this.keyWords, totalChunks, this.flags);
            }
            // Fill the current chunk
            const want = constants_js_1.CHUNK_LEN - this.chunkState.len();
            const take = Math.min(want, inputLen);
            this.chunkState.update(input, inputOffset, take);
            inputOffset += take;
            inputLen -= take;
        }
        return this;
    }
    /**
     * Get the output parameters (for XOF mode or finalization).
     */
    finalizeOutput() {
        let output = this.chunkState.output();
        let parentBlock = this.parentBlock;
        let cv = this.finalizeCv;
        // If there are chunks on the stack, merge them
        if (this.cvStackLen > 0) {
            // First compress the current chunk
            (0, compress_js_1.compress)(output.inputCv, 0, output.blockWords, 0, cv, 0, false, output.counter, output.blockLen, output.flags);
            // Merge with parent nodes from stack
            while (this.cvStackLen > 0) {
                this.cvStackLen--;
                parentBlock.set(this.cvStack.subarray(this.cvStackLen * 8, (this.cvStackLen + 1) * 8), 0);
                parentBlock.set(cv, 8);
                if (this.cvStackLen > 0) {
                    (0, compress_js_1.compress)(this.keyWords, 0, parentBlock, 0, cv, 0, false, 0, constants_js_1.BLOCK_LEN, this.flags | constants_js_1.PARENT);
                }
                else {
                    // This is the root - return output params
                    return {
                        inputCv: this.keyWords,
                        blockWords: parentBlock,
                        blockLen: constants_js_1.BLOCK_LEN,
                        counter: 0,
                        flags: this.flags | constants_js_1.PARENT,
                    };
                }
            }
        }
        // Single chunk case
        return output;
    }
    /**
     * Finalize the hash and return the result.
     *
     * @param outputLength - Number of bytes to output (default: 32)
     * @returns The hash output
     */
    finalize(outputLength = constants_js_1.OUT_LEN) {
        const output = this.finalizeOutput();
        const result = new Uint8Array(outputLength);
        if (outputLength <= 64) {
            const outWords = this.outWords;
            (0, compress_js_1.compress)(output.inputCv, 0, output.blockWords, 0, outWords, 0, outputLength > 32, // full output if > 32 bytes
            output.counter, output.blockLen, output.flags | constants_js_1.ROOT);
            if (utils_js_1.IS_LITTLE_ENDIAN) {
                const outBytes = new Uint8Array(outWords.buffer);
                result.set(outBytes.subarray(0, outputLength));
            }
            else {
                (0, utils_js_1.writeLittleEndianBytesPartial)(outWords, 0, result, 0, outputLength);
            }
        }
        else {
            // Multiple blocks - use XOF
            const xof = this.finalizeXof();
            const full = xof.read(outputLength);
            result.set(full);
        }
        return result;
    }
    /**
     * Finalize and return an XOF reader for arbitrary-length output.
     */
    finalizeXof() {
        const output = this.finalizeOutput();
        return new XofReader(new Uint32Array(output.inputCv), new Uint32Array(output.blockWords), output.counter, output.blockLen, output.flags);
    }
}
exports.Hasher = Hasher;
