// @huggingface/blake3-jit/dist/commonjs/index.js
"use strict";
/**
 * BLAKE3 - The fastest pure JavaScript implementation
 *
 * Features:
 * - All 3 modes: hash, keyed (MAC), derive_key
 * - XOF (eXtendable Output Function) support
 * - Automatic WASM SIMD acceleration for large inputs
 * - Zero dependencies
 * - Tree-shakeable exports
 *
 * @example
 * ```typescript
 * import { hash, createKeyed, createDeriveKey } from 'blake3-jit';
 *
 * // Simple hashing
 * const digest = hash(new Uint8Array([1, 2, 3]));
 *
 * // Keyed hashing (MAC)
 * const mac = createKeyed(key).update(data).finalize();
 *
 * // Key derivation
 * const derived = createDeriveKey("my context").update(material).finalize(64);
 * ```
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.warmupSimd = exports.hashInto = exports.hash = exports.XofReader = exports.Hasher = void 0;
exports.createKeyed = createKeyed;
exports.createDeriveKey = createDeriveKey;
exports.createHasher = createHasher;
// Core exports
var hasher_js_1 = require("./hasher.js");
Object.defineProperty(exports, "Hasher", { enumerable: true, get: function () { return hasher_js_1.Hasher; } });
Object.defineProperty(exports, "XofReader", { enumerable: true, get: function () { return hasher_js_1.XofReader; } });
var hash_js_1 = require("./hash.js");
Object.defineProperty(exports, "hash", { enumerable: true, get: function () { return hash_js_1.hash; } });
Object.defineProperty(exports, "hashInto", { enumerable: true, get: function () { return hash_js_1.hashInto; } });
Object.defineProperty(exports, "warmupSimd", { enumerable: true, get: function () { return hash_js_1.warmupSimd; } });
// Convenience imports
const hasher_js_2 = require("./hasher.js");
/**
 * Create a new keyed hasher (MAC).
 *
 * @param key - 32-byte key
 * @returns A new Hasher configured for keyed hashing
 *
 * @example
 * ```typescript
 * const key = new Uint8Array(32); // Your 32-byte key
 * crypto.getRandomValues(key);
 *
 * const mac = createKeyed(key)
 *   .update(message)
 *   .finalize();
 * ```
 */
function createKeyed(key) {
    return hasher_js_2.Hasher.newKeyed(key);
}
/**
 * Create a new key derivation hasher.
 *
 * @param context - Context string for domain separation
 * @returns A new Hasher configured for key derivation
 *
 * @example
 * ```typescript
 * const derivedKey = createDeriveKey("my-app encryption key v1")
 *   .update(inputKeyMaterial)
 *   .finalize(32);
 * ```
 */
function createDeriveKey(context) {
    return hasher_js_2.Hasher.newDeriveKey(context);
}
/**
 * Create a new regular hasher for incremental hashing.
 *
 * @returns A new Hasher
 *
 * @example
 * ```typescript
 * const hasher = createHasher();
 * hasher.update(chunk1);
 * hasher.update(chunk2);
 * const digest = hasher.finalize();
 * ```
 */
function createHasher() {
    return new hasher_js_2.Hasher();
}
// Import for default export
const hash_js_2 = require("./hash.js");
// Pre-warm SIMD in browser environments (non-blocking)
// This avoids initialization latency on first large hash
if (typeof globalThis !== "undefined" && typeof globalThis.document !== "undefined") {
    queueMicrotask(() => {
        (0, hash_js_2.warmupSimd)();
    });
}
// Default export for convenience
exports.default = {
    hash: hash_js_2.hash,
    hashInto: hash_js_2.hashInto,
    Hasher: hasher_js_2.Hasher,
    createHasher,
    createKeyed,
    createDeriveKey,
};
